### PR TITLE
HostVolumes: Include volume info in nomad node status

### PR DESCRIPTION
**Pull request by endocrimes**
NOTE: When #5923 is merged into `f-host-volumes`, this PR should be targeted against `f-host-volumes`

## Short Output

```
$ nomad node status -self -short
ID           = 964f7030
Name         = mew
Class        = <none>
DC           = dc1
Drain        = false
Eligibility  = eligible
Status       = ready
Host Volumes = tmp-dir
Drivers      = docker,mock_driver,raw_exec

Allocations
No allocations placed
```

## Verbose Output
```
ID          = 964f7030-9882-a359-7965-2bd432fbdd96
Name        = mew
Class       = <none>
DC          = dc1
Drain       = false
Eligibility = eligible
Status      = ready
Uptime      = 168h8m18s

Host Volumes
Name     ReadOnly  Hidden  Source
tmp-dir  false     false   /tmp

Drivers
Driver       Detected  Healthy  Message                  Time
docker       true      true     Healthy                  2019-07-25T15:30:20+02:00
exec         false     false    Driver must run as root  2019-07-25T15:30:20+02:00
java         false     false    Driver must run as root  2019-07-25T15:30:20+02:00
mock_driver  true      true     Healthy                  2019-07-25T15:30:20+02:00
qemu         false     false    <none>                   2019-07-25T15:30:20+02:00
raw_exec     true      true     Healthy                  2019-07-25T15:30:20+02:00
rkt          false     false    Driver must run as root  2019-07-25T15:30:20+02:00

Node Events
Time                       Subsystem  Message          Details
2019-07-25T15:30:20+02:00  Cluster    Node registered  <none>

Allocated Resources
CPU          Memory      Disk
0/32000 MHz  0 B/15 GiB  0 B/7.5 GiB

Allocation Resource Utilization
CPU          Memory
0/32000 MHz  0 B/15 GiB

Host Resource Utilization
CPU            Memory         Disk
516/32000 MHz  11 GiB/15 GiB  (tmpfs)

Allocations
No allocations placed

Attributes
{{omitted}}

Meta
```

### CODE REVIEW

File: command/node_status.go [hunk 1]
```diff
@@ -347,6 +358,11 @@ func (c *NodeStatusCommand) formatNode(client *api.Client, node *api.Node) int {
 			basic = append(basic, fmt.Sprintf("Uptime|%s", uptime.String()))
 		}
 
+		// Emit the volume info
+		if !c.verbose {
```
Comment (notnoop) [inline command/node_status.go]: It took me a second to understand the logic here.  The comment (here and driver info) isn't informative and misleading.  Maybe more "Emit volume info summary; when in verbose, emit detail info below?" or something like would have helped my reading.

File: command/node_status.go [hunk 2]
```diff
@@ -443,6 +460,27 @@ func (c *NodeStatusCommand) outputTruncatedNodeDriverInfo(node *api.Node) string
...
 
+func (c *NodeStatusCommand) outputNodeVolumeInfo(node *api.Node) {
+	c.Ui.Output(c.Colorize().Color("\n[bold]Host Volumes"))
+
+	size := len(node.HostVolumes)
+	nodeVolumes := make([]string, 0, size+1)
+
+	nodeVolumes = append(nodeVolumes, "Name|ReadOnly|Hidden|Source")
+
+	volumes := make([]string, 0, len(node.HostVolumes))
```
Comment (notnoop) [inline command/node_status.go]: I find the use of generic names here (e.g. `volumes` and `nodeVolumes` and `node.HostVolumes`) all for different concepts to make it a bit hard to follow.  Consider changing `volumes` to `volumeNames`, and `nodeVolumes` to `volumeInfo`.  Given that this function is all in context of volumes, I may even suggest dropping `volume` prefix.

Another thing that confused me was the transition from operating on `nodeVolumes` to operating on `volumes` then back to `nodeVolumes`.  I would suggest keeping all nodeVolumes work together so it's easier to understand its role and logic without the names detour.

File: command/node_status.go [hunk 3]
```diff
@@ -299,6 +299,16 @@ func nodeDrivers(n *api.Node) []string {
 	return drivers
 }
 
+func nodeVolumes(n *api.Node) []string {
```
Comment (notnoop) [inline command/node_status.go]: Maybe call it `nodeVolumeNames`?  also I liked how `outputNodeVolumeInfo` used `name` as the variable in the loop.

File: command/node_status.go [hunk 4]
```diff
@@ -333,6 +343,7 @@ func (c *NodeStatusCommand) formatNode(client *api.Client, node *api.Node) int {
 	}
 
 	if c.short {
+		basic = append(basic, fmt.Sprintf("Host Volumes|%s", strings.Join(nodeVolumes(node), ",")))
```
Comment (notnoop) [inline command/node_status.go]: Do we want to emit a line here if there is no node host volumes?

File: command/node_status.go (hunk 1 above)
Comment (endocrimes) [reply command/node_status.go]: Yeah this definitely isn't my favourite code either - I was mostly copying the existing style here, to avoid doing a bigger cleanup. But I guess I'll do that when I come back to this

File: command/node_status.go (hunk 4 above)
Comment (endocrimes) [reply command/node_status.go]: This is maintaining the same behavior as for drivers right now - Omitting it entirely is probably more confusing than an empty list imo?

File: command/node_status.go (hunk 3 above)
Comment (endocrimes) [reply command/node_status.go]: Maintaining the same style as `nodeDrivers` for this. We can do a bigger refactor of this command at some point, but this probably isn't the right place for it.

File: command/node_status.go (hunk 4 above)
Comment (notnoop) [reply command/node_status.go]: I don't know what users expect.  Drivers field is almost never empty (otherwise, what's the point of clients), while host volumes is a feature that will take awhile to gain adoption

File: command/node_status.go (hunk 3 above)
Comment (notnoop) [reply command/node_status.go]: I would suggest for going for better clarity in new code.  Matchingold code ambiguity and tech debt feels like an odd justification for writing non-ideal new code tbqh.

File: command/node_status.go (hunk 4 above)
Comment (endocrimes) [reply command/node_status.go]: Not showing them at all would be confusing imo, bc it could fairly easily lead to the confusion of e.g "oh am i running the wrong cli version" etc? - And in the format of output we use, I'm not sure there's really an "empty" entry that isn't an empty list that would make sense?

Comment (notnoop) [reply command/node_status.go]: makes sense - either way, nbd.

**Comment by endocrimes** [comment]
@notnoop Generally cleaned up a couple of parts of this command to make it a bit easier to understand what's going on.

- Made short be more self contained and return early, so we can not think about it for the remainder of the formatter.
- Improved comments around what we're rendering

File: command/node_status.go [hunk 5]
```diff
@@ -333,86 +343,109 @@ func (c *NodeStatusCommand) formatNode(client *api.Client, node *api.Node) int {
...
 		return 1
 	}
 
+	return 0
+}
+
+func (c *NodeStatusCommand) outputAllocInfo(client *api.Client, node *api.Node) error {
+	nodeAllocs, _, err := client.Nodes().Allocations(node.ID, nil)
+	if err != nil {
+		return err
```
Comment (notnoop) [inline command/node_status.go]: I'd return the error prefix here instead of the caller side.  Had the function been `queryAllocInfo`, having callers add the prefix would be reasonable imo.

```suggestion
		return fmt.Errorf("failed to query node allocations: %s", err)
```

File: command/node_status.go [hunk 6]
```diff
@@ -333,86 +343,109 @@ func (c *NodeStatusCommand) formatNode(client *api.Client, node *api.Node) int {
...
-			driverStatus := fmt.Sprintf("Driver Status| %s", c.outputTruncatedNodeDriverInfo(node))
-			basic = append(basic, driverStatus)
+		// Output alloc info
+		if err := c.outputAllocInfo(client, node); err != nil {
+			c.Ui.Error(fmt.Sprintf("Error querying node allocations: %s", err))
+			return 1
 		}
 
-		c.Ui.Output(c.Colorize().Color(formatKV(basic)))
+		return 0
```
Comment (notnoop) [inline command/node_status.go]: Terminating early here is very nice; it's much more obvious what the `short` flag displays here imo - thanks for doing this!
